src/lib.rs

```diff
@@ -107,3 +118,46 @@ unsafe fn secbuf_desc(bufs: &mut [sspi::SecBuffer]) -> sspi::SecBufferDesc {
         pBuffers: bufs.as_mut_ptr(),
     }
 }
+
+fn alpn_list(protos: &[Vec<u8>]) -> Vec<u8> {
+    // ALPN wire format is each ALPN preceded by its length as a byte.
+    let mut alpn_wire_format = Vec::with_capacity(
+        protos.iter().map(Vec::len).sum::<usize>() + protos.len(),
+    );
+    for alpn in protos {
+        alpn_wire_format.push(alpn.len() as u8);
+        alpn_wire_format.extend(alpn);
+    }
+
+    // Make sure that the memory we're using for `sspi::SEC_APPLICATION_PROTOCOLS` matches
+    // its current 4-byte alignment requirements.
+    let size = SEC_APPLICATION_PROTOCOL_HEADER_SIZE + alpn_wire_format.len();
+    let mut aligned = Vec::<u32>::with_capacity((size - 1) / mem::size_of::<u32>() + 1);
+    let p = aligned.as_mut_ptr() as *mut u8;
+    let cap = aligned.capacity() * (mem::size_of::<u32>() / mem::size_of::<u8>());
+
+    mem::forget(aligned);
+
+    unsafe {
+        let mut buf = Vec::from_raw_parts(p, 0, cap);
+        buf.resize(size, 0);
+
+        let protocols = &mut *(buf.as_mut_ptr() as *mut sspi::SEC_APPLICATION_PROTOCOLS);
+        protocols.ProtocolListsSize =
+            (SEC_APPLICATION_PROTOCOL_LIST_HEADER_SIZE + alpn_wire_format.len()) as ctypes::c_ulong;
+
+        let protocol = &mut *protocols.ProtocolLists.as_mut_ptr();
+        protocol.ProtoNegoExt = sspi::SecApplicationProtocolNegotiationExt_ALPN;
+        protocol.ProtocolListSize = alpn_wire_format.len() as ctypes::c_ushort;
+
+        let protocol_list_start = protocol.ProtocolList.as_mut_ptr();
+        let struct_end = buf.as_ptr().add(size);
```

Comment:
You should be able to just use, which should be better readable:
```
let protocol_list_offset = protocol.ProtocolList.as_ptr() as usize - buf.as_ptr() as usize;
let protocol_list = &mut buf[protocol_list_offset..];
protocol_list.copy_from_slice(&alpn_wire_format);
```

Reply:
Oh good point, sounds good.